Re-implement side-tracing.

ptersilie opened the pull request:
This PR fixes side-tracing which we disabled when we replaced the traceinputs struct with passing live variables directly via registers/stack offsets.

In addition this also implements side-trace looping, i.e. when a side-trace reaches the end without any guards failing we no longer deoptimise back to the interpreter but instead loop back to the root trace.

Getting everything working was a very messy ordeal and while I tried to keep commits consistent I may not always have succeeded. Before merging we'll need to squash this down into a single commit. I don't know if the commits actually help understanding this better. My suggestion would be to review all changes at once and only look at the individual commits if it's unclear why we are doing things this way (the commit messages may give some insight). But I'll leave that choice up to you. Good luck!

Here's some performance data. Some benchmarks give good speedups (Permute, Storage, Queens, Sieve), others have slowdowns (Bounce, Nbody).

| Benchmark  | Sidetracing | Normal | Speedup |
|------------|-------------|--------|---------|
| Richards   | 14421       | 13975  | 0.97    |
| Json       | 3115        | 3069   | 0.99    |
| Permute    | 4270        | 5213   | 1.22    |
| Storage    | 5182        | 6878   | 1.33    |
| Queens     | 2969        | 4418   | 1.49    |
| DeltaBlue  | 2312        | 2210   | 0.96    |
| Havlak     | 13792       | 14534  | 1.05    |
| List       | 3267        | 3191   | 0.98    |
| Bounce     | 9226        | 6497   | 0.70    |
| CD         | 8026        | 7997   | 1.00    |
| NBody      | 7557        | 3365   | 0.45    |
| Sieve      | 3922        | 8652   | 2.21    |
| Mandelbrot | 1517        | 1514   | 1.00    |
| Towers     | 4524        | 4356   | 0.96    |

ltratt commented:
Do the benchmark numbers have synchronised compilation on or off?

ptersilie commented:
> Do the benchmark numbers have synchronised compilation on or off?

Synchronized is and release mode.

ltratt commented on ykrt/src/compile/guard.rs:
```diff
@@ -35,10 +35,9 @@ impl Guard {
                     true
                 } else {
-                    // FIXME: temporarily disable side-tracing by not ever adding to the `failed` count.
                     #[allow(clippy::identity_op)]
```
You can now remove the `#[allow]` and the brackets.

ltratt commented on ykrt/src/compile/jitc_yk/codegen/mod.rs:
```diff
@@ -24,6 +24,7 @@ pub(crate) trait CodeGen: Send + Sync {
         m: Module,
         mt: Arc<MT>,
         hl: Arc<Mutex<HotLocation>>,
+        sp_offset: Option<usize>,
```
What is `sp_offset`?

ptersilie replied:
There's a doc string on the implementation of codegen. Maybe we should move all of the doc strings over to the trait instead?

ltratt replied:
I'm not going to pretend I understand that sentence :p

ptersilie replied:
The docstring is here: https://github.com/ykjit/yk/pull/1422/files#diff-2d8235b02f7af5cbc0cef1cf9b9789aaf37ee8bee882f4bd0f9c266e1d142edbR61

I feel like it's more relevant there. But I can copy it over if you insist.

ltratt replied:
You can reference it with rustdoc syntax, that way it can't get out of sync.

ptersilie replied:
And there's another here: https://github.com/ykjit/yk/pull/1422/files#diff-b7513e8c31bfe56ee2b41fa041e8632d81f4bafc2e759b22b597f3867359269eR156

ltratt replied:
Right, but when a dummy like me reads local code, they won't know this global context. rustdoc links (which editors often understand) are one solution to this.

ptersilie replied:
The problem with this design is that we are often passing values through various files. Do we always copy the docstrings to every single occurence? I don't know what a good standard for that is. e.g. do we also comment `m, mt, hl`?

ptersilie replied:
For example. Should we also comment `sp_offset` here:
- https://github.com/ykjit/yk/pull/1422/files#diff-b7513e8c31bfe56ee2b41fa041e8632d81f4bafc2e759b22b597f3867359269eR124
- https://github.com/ykjit/yk/pull/1422/files#diff-b7513e8c31bfe56ee2b41fa041e8632d81f4bafc2e759b22b597f3867359269eR162

The second of which seems even more important than the trait you commented on which we are hardly going to look at very often.

ptersilie replied:
@ltratt I've tried to improve the `sp_offset` doc strings. Sadly, Rust doesn't allow us to use doc strings for function parameters so I've used normal comments instead. But I've used doc strings wherever possible: [6589d14](https://github.com/ykjit/yk/pull/1422/commits/6589d14b0dfb171bd513fba0aebd78c3241476e4)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs:
```diff
@@ -227,6 +227,10 @@ impl<'a> LSRegAlloc<'a> {
         self.stack.size()
     }
 
+    pub(crate) fn stack_size(&mut self) -> usize {
```
Add a doc string (even I can't remember if this is bytes or #entries, and I wrote the code I think!).

ptersilie replied:
[0022632](https://github.com/ykjit/yk/pull/1422/commits/00226328b8a1cfff73af85e11612135268003639)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs:
```diff
@@ -269,6 +273,10 @@ impl LSRegAlloc<'_> {
         self.spills[usize::from(iidx)] = SpillState::Stack(frame_off);
     }
 
+    pub(crate) fn assign_const(&mut self, iidx: InstIdx, bits: u32, v: u64) {
```
Add a doc string. [Should be easy to copy from the other similar functions.] I also think this should be `assign_const_int` because it can't (yet) handle other `Const` kinds.

ptersilie replied:
[0022632](https://github.com/ykjit/yk/pull/1422/commits/00226328b8a1cfff73af85e11612135268003639)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -148,55 +149,92 @@ struct Assemble<'a> {
+    /// interpreter frame itself and all parent traces. Since all traces execute in the interpreter
+    /// frame, this trace adds to this value, making extra space on the stack. This then forms the
+    /// new `sp_offset` for any side-traces of this trace.
+    sp_offset: usize,
```
Should we make this a new type? Then we would always know what it's doing.

ptersilie replied:
I'm unsure it's worth the extra hassle of having to convert between types every time we use it.

ltratt replied:
In general, I think one can overuse newtypes, but this might be one case where it's worth doing. But, fair enough, we can do it in another PR.

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -758,9 +830,9 @@ impl<'a> Assemble<'a> {
-                // The third argument to the control point is the stackmap ID and is passed by RDX.
-                // We don't anticipate this ever being a live variable.
-                unreachable!()
+                // While we would never see this case in normal traces (since the control point
```
I think we can just say "This case only happens in side-traces" or something equally simple.

ptersilie replied:
[6ca1b84](https://github.com/ykjit/yk/pull/1422/commits/6ca1b8481135aacac479f697b24350d98ff83886)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -1659,22 +1902,55 @@ pub(super) struct X64CompiledTrace {
-        let callframes = self.deoptinfo[&usize::from(gidx)].inlined_frames.clone();
+        let callframes = deoptinfo.inlined_frames.clone();
+
+        // Get the root trace from the HotLocation
```
I think this should have `FIXME: It might be better if we had a consistent mechanism for doing this rather than fishing it out of the `HotLocationKind` or similar.

ptersilie replied:
[1e69032](https://github.com/ykjit/yk/pull/1422/commits/1e69032bcb639fd3c88fc00fab48075378743169)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -1659,22 +1902,55 @@ pub(super) struct X64CompiledTrace {
+                .as_any()
+                .downcast::<X64CompiledTrace>()
+                .unwrap(),
+            // Since we are currently trying to compile a side-trace, the current HotLocationKind
```
This comment is wrong, as can be seen from the fact that there are two cases in the `match`!

ptersilie replied:
[1e69032](https://github.com/ykjit/yk/pull/1422/commits/1e69032bcb639fd3c88fc00fab48075378743169)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -1659,22 +1902,55 @@ pub(super) struct X64CompiledTrace {
+            lives,
             callframes,
+            // FIXME: This needs to point to the root trace, not the immediate parent.
+            parent: ParentTracePtr(root_ctr.entry()), //ParentTracePtr(self.entry()),
```
Commented code.

Also this `FIXME` sounds unusually bad?!

ptersilie replied:
The comment is old. I just need to remove it.

ptersilie replied:
[868cf76](https://github.com/ykjit/yk/pull/1422/commits/868cf769fe2bf7de7725d55f99422731ca074d4c)

ltratt commented on ykrt/src/compile/jitc_yk/jit_ir/mod.rs:
```diff
@@ -562,14 +567,35 @@ impl Module {
         self.loop_start_vars.push(op);
     }
 
+    pub(crate) fn set_parent_entry_vars(&mut self, entry_vars: &[VarLocation]) {
```
Add doc strings for all new functions please.

ptersilie replied:
[9a39a8b](https://github.com/ykjit/yk/pull/1422/commits/9a39a8bb1c45f9128f50360f7dfa1391c4da6516)

ltratt commented:
Comments:

* We need C tests for this, even if they're very basic. 
* Do we need `parent_jump` to have such a specific name? Isn't the point that it's just `jump` to the beginning of the main trace?

ltratt commented on ykrt/src/compile/jitc_yk/trace_builder.rs:
```diff
@@ -94,51 +91,49 @@ impl TraceBuilder {
+
+            // Get the location for this input variable.
+            let var = &rec.live_vars[idx];
+            if var.len() > 1 {
```
Is this an `assert` in disguise or a real `todo`?

ptersilie replied:
[9932886](https://github.com/ykjit/yk/pull/1422/commits/9932886c1f495981805e07b771bd4a1290defc9a)

ltratt commented on ykrt/src/compile/jitc_yk/trace_builder.rs:
```diff
@@ -1125,39 +1126,59 @@ impl TraceBuilder {
                 })
                 .collect::<Vec<_>>();
+
+            //#[cfg(tracer_swt)]
```
This commented code looks fishy.

ptersilie replied:
Whoops. That definitely needs to go back in. I commented it out to get rust-analyser to give me hints.

ptersilie replied:
[b5a496d](https://github.com/ykjit/yk/pull/1422/commits/b5a496d546cf2f1eee21b143b60bf77ece9bbb59)

ptersilie commented:
Addressed comments. Also note the commit [8189ad0](https://github.com/ykjit/yk/pull/1422/commits/8189ad054040160746a3e7f02e3d2650c297093d) which simplifies the new `ParentJump` instruction.

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/deopt.rs:
```diff
@@ -41,8 +41,8 @@ pub(crate) extern "C" fn __yk_guardcheck(gidx: u64) -> *const libc::c_void {
         let staddr = st.entry();
-        let mt = Arc::clone(&ctr.mt);
-        mt.log.log(Verbosity::JITEvent, "execute-side-trace");
+        //let mt = Arc::clone(&ctr.mt);
```
Eh?

ptersilie replied:
Oh, I wanted to `git absorb` this but I must have commited it accidentally. Yeah, it's put back in https://github.com/ykjit/yk/pull/1422/commits/0a3d0820e5bc843655f9e6236b5a82d2b27dbf83. Sorry.

ltratt commented:
A few minor things, but we're almost there.

ptersilie commented:
>     * We need C tests for this, even if they're very basic.

Sorry, forgot to add them. See [abc8697](https://github.com/ykjit/yk/pull/1422/commits/abc8697a0a45c490c855867ebba444191b9c7882)

>     * Do we need `parent_jump` to have such a specific name? Isn't the point that it's just `jump` to the beginning of the main trace?

It's different to a normal trace loop jump, so I think it needs another name. But I've renamed it to `RootJump` which is more accurate.

ltratt commented on ykrt/src/compile/jitc_yk/codegen/mod.rs:
```diff
@@ -24,6 +24,8 @@ pub(crate) trait CodeGen: Send + Sync {
         m: Module,
         mt: Arc<MT>,
         hl: Arc<Mutex<HotLocation>>,
+        // Stack pointer offset from the base pointer of interpreter frame as defined in
```
This needs to be `///`.

ptersilie replied:
Rust doesn't allow this: https://github.com/ykjit/yk/pull/1422#discussion_r1804388706

ltratt replied:
Sorry, I misread the diff.

The standard way to do this is to put this in the function's docstring, not as a comment in the parameter list.

ptersilie replied:
[0d16998](https://github.com/ykjit/yk/pull/1422/commits/0d16998eeb2f3b539bf7169dd057680224781d55)

ltratt commented on ykrt/src/compile/jitc_yk/codegen/x64/mod.rs:
```diff
@@ -121,6 +121,8 @@ impl CodeGen for X64CodeGen {
         mt: Arc<MT>,
         hl: Arc<Mutex<HotLocation>>,
+        // Stack pointer offset from the base pointer of interpreter frame as defined in
+        // [YkSideTraceInfo].
```
I think this should be `[YkSideTraceInfo::sp_offset]` or similar.

ptersilie replied:
Here you go: [293a9b5](https://github.com/ykjit/yk/pull/1422/commits/293a9b518a5cf75c82c5abeaf33b9dacc1febc13)

ltratt commented:
Please squash.

ptersilie commented:
Squashed.
